### ISSUE TITLE
[BUG] Severe Performance Degradation When Using `pennylane_qiskit` with Torch Training Loop

### ISSUE THREAD

**Issue by henryzou50** [PennyLaneAI/pennylane#7973]
### Expected behavior

Running a hybrid QML model using a Torch-based training loop with `pennylane_qiskit` and Qiskit backends (e.g., `qiskit.aer` or `qiskit.remote`) should have comparable performance to using `default.qubit`. We expect that:

- Circuit execution should be efficiently batched or reused during training.
- Training time should remain reasonable and not scale disproportionately with the number of mini-batches.
- The plugin should support session reuse or efficient submission of circuits per batch, especially when used with `TorchLayer`.

### Actual behavior

When switching from `default.qubit` to a Qiskit backend (`qiskit.aer` or `qiskit.remote`), the training time increases by approximately 100x. Each mini-batch appears to trigger a full job submission, resulting in extremely slow performance.

This slowdown makes the plugin impractical for Torch training workflows, especially when integrated into hybrid neural networks using `TorchLayer`.

The recent PR for session support (`#140`) does not appear to improve performance in this case.

### Additional information

Example device setup:
```python
# Fast (expected behavior)
dev = qml.device('default.qubit', wires=5)

# Slow (problematic behavior)
dev = qml.device('qiskit.aer', wires=5, backend='unitary_simulator')
dev = qml.device('qiskit.remote', wires=5, backend=backend)
```

The model is a hybrid Torch model using `qml.qnn.TorchLayer`. Each epoch contains many mini-batches, and the slowdown suggests that sessions or circuits are not efficiently managed under the hood.

### Source code

```shell

```

### Tracebacks

```shell

```

### System information

```shell
Name: pennylane
Version: 0.42.1
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page: 
Author: 
Author-email: 
License-Expression: Apache-2.0
Requires: appdirs, autograd, autoray, cachetools, diastatic-malt, networkx, numpy, packaging, pennylane-lightning, requests, rustworkx, scipy, tomlkit, typing_extensions
Required-by: pennylane_lightning

Platform info:           macOS-14.6.1-arm64-arm-64bit
Python version:          3.11.4
Numpy version:           2.3.2
Scipy version:           1.16.1
Installed devices:
- lightning.qubit (pennylane_lightning-0.42.0)
- default.clifford (pennylane-0.42.1)
- default.gaussian (pennylane-0.42.1)
- default.mixed (pennylane-0.42.1)
- default.qubit (pennylane-0.42.1)
- default.qutrit (pennylane-0.42.1)
- default.qutrit.mixed (pennylane-0.42.1)
- default.tensor (pennylane-0.42.1)
- null.qubit (pennylane-0.42.1)
- reference.qubit (pennylane-0.42.1)
```

### Existing GitHub issues

- [x] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by mlxd** [comment]
Hi @henryzou50 

Based on your above results, this behaviour seems expected to me.

`default.qubit` as a device is transparent to the ML interface as all representations use the named Torch, Jax and TF data to represent the statevector, results and operators. This allows the use of batching/broadcasting, as well as native use with backpropagation, since the visibility of the internal representations are completely transparent.

For other devices this integration is performed using the ML interface's wrappings. This requires an opaque method to wrap the device I/O which takes inputs and outputs in the ML interface type, but must natively convert it and bind to the device's execution methods. As a result, `default.qubit` should be much faster in this case, as it supports:

* Native backpropagation through the device.
* As an alternative, adjoint gradients are also supported too.
* Batching works directly, due to the transparent design.

Other devices either must provide a custom execution/gradient implementation, or they will fall-back to quantum hardware native parameter-shift implementation, which involves a lot more executions for evaluation of the gradients.

Given you are using small numbers of qubits/wires, `default.qubit` is likely the best candidate for your use-case. If you want hardware execution, the above points will likely be a bottleneck.


If you think there's somewhere in the execution pipeline for the Qiskit backend that can be improved, we'd be more than happy to receive a submission [here](https://github.com/PennyLaneAI/pennylane-qiskit/) to resolve the above points.

Otherwise, if you have a minimum working example workload that allows us to see the issue and perform a more detailed analysis to validate the above, we'd also be happy to take a look.

**Comment by henryzou50** [comment]
Thanks for the quick response. Here's a minimal working example provided by the user that reproduces the performance issue. You can uncomment the `qiskit.aer` device line to observe the slowdown compared to `default.qubit`.

Adding the script below for reference:
```python


import tensorflow as tf
import torch
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import confusion_matrix
import argparse
import os
import random
import pennylane as qp
from timeit import default_timer as timer

from qiskit_ibm_runtime.fake_provider import FakeLimaV2
backend =FakeLimaV2()

def set_seeds(s):
    tf.random.set_seed(s)
    torch.manual_seed(s)
    np.random.seed(s)
    random.seed(s)
    os.environ['PYTHONHASHSEED'] = str(s)

n_qubits=5
# 1 qubit per feature
#dev = qp.device('qiskit.remote', wires=n_qubits,backend=backend)
#dev = qp.device('qiskit.aer',wires=n_qubits,backend='unitary_simulator')

dev = qp.device('default.qubit',wires=n_qubits)    

class QML:
    def __init__(self, model_id, run_number):
        self.model_id = model_id
        self.run_number = run_number
        if self.model_id == 'simple_hybrid_pqc':
            self.model = self._simple_hybrid_model()
        else:
            self.model = self._simple_nn_model()


    def _simple_hybrid_model(self):
     
        @qp.qnode(dev,interface='torch')
        def model_circuit(inputs,params):
            inputs = 2*inputs
            qp.AngleEmbedding(inputs, wires=range(n_qubits))
            qp.StronglyEntanglingLayers(weights=params, wires=range(n_qubits))
            return [qp.expval(qp.PauliZ(i)) for i in range(n_qubits)]
        
        params_shape ={"params": (3,5,3)}
        circuit_layer = qp.qnn.TorchLayer(model_circuit,params_shape)
        clayer_1 = torch.nn.ELU(10)
        clayer_2 = torch.nn.Linear(5,3,dtype=float)
        clayer_3 = torch.nn.Softmax(dim=1)
        layers = [circuit_layer,clayer_1,clayer_2,clayer_3]
        model = torch.nn.Sequential(*layers)
        return model

    def _simple_nn_model(self):
        model = tf.keras.Sequential([tf.keras.layers.Dense(10, activation='elu'),
                                     tf.keras.layers.Dense(3, activation='softmax')])
        return model

    def prepare_classic_dataset(self, data):
        x = data['x']
        y = data['y']
        x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20)
        scaler = MinMaxScaler(feature_range=(0, np.pi))
        x_train = scaler.fit_transform(x_train)
        x_test = scaler.transform(x_test)
        y_train_hot = np.eye(3)[y_train]
        y_test_hot = np.eye(3)[y_test]

        return x_train, x_test, y_train,y_test, y_train_hot, y_test_hot

    def prepare_quantum_dataset(self, x_train,x_test):
        x_train_q = torch.tensor(x_train)
        x_test_q = torch.tensor(x_test)

        return x_train_q, x_test_q

    def train(self, x_train, y_train, data_loader,learning_rate, epochs,batch_size):
        optimizer = torch.optim.Adam(self.model.parameters(),lr=learning_rate)
        loss = torch.nn.CrossEntropyLoss()
        history = []
        accuracy = []

        batches = len(x_train) // batch_size

        for epoch in range(epochs):
            running_loss = 0
            
            for xs,ys in data_loader:
                optimizer.zero_grad()
                loss_evaluated = loss(self.model(xs),ys)
                loss_evaluated.backward()
                optimizer.step()
                running_loss += loss_evaluated
                
            print(epoch)    
            avg_loss = running_loss / batches
            history.append(avg_loss)
            
            y_pred = self.model(x_train)
            predictions = torch.argmax(y_pred, axis=1).detach().numpy()
            correct = [1 if p == p_true else 0 for p, p_true in zip(predictions, y_train)]
            accuracy.append(sum(correct) / len(correct))  
        return history, accuracy

    def evaluate(self, x_train, x_test, y_train, y_test):
        _ = self.model(x_train)

        y_pred = self.model(x_train)
        predictions = torch.argmax(y_pred, axis=1).detach().numpy()
        correct = [1 if p == p_true else 0 for p, p_true in zip(predictions, y_train)]
        accuracy_train = sum(correct) / len(correct)
        
        loss = torch.nn.CrossEntropyLoss()
        loss_train = loss(m.model(x_train),torch.tensor(y_train))
        
        y_test_pred = self.model(x_test)       
        predictions = torch.argmax(y_test_pred, axis=1).detach().numpy()
        correct = [1 if p == p_true else 0 for p, p_true in zip(predictions, y_test)]
        accuracy_test = sum(correct) / len(correct)       
        loss_test = loss(m.model(x_test),torch.tensor(y_test))
        

        # confusion matrix
        y_test_ne = np.argmax(y_test, axis=1)
        y_test_pred_ne = np.argmax(y_test_pred.detach().numpy(), axis=1)
        cm = confusion_matrix(y_test_ne, y_test_pred_ne)

        return {'train_accuracy': accuracy_train,
                'test_accuracy': accuracy_test,
                'train_loss': loss_train,
                'test_loss': loss_test,
                'cm': cm,
                'run': self.run_number}
# argparser!
parser = argparse.ArgumentParser(description='Hyperparameters for the experiment')
parser.add_argument('--model_type', action="store", dest="model_type", type=str, default='simple_hybrid_pqc')
parser.add_argument('--learning_rate', action="store", dest="learning_rate", type=float, default=0.01)
parser.add_argument('--epochs', action="store", dest="epochs", type=int, default=200)
parser.add_argument('--n_runs', action="store", dest="n_runs", type=int, default=1)
args = parser.parse_args()

if __name__ == "__main__":

    results = []
    seeds = open("seeds.txt", "r").readlines()
    batch_size = 32
    start = timer()

    for n in range(args.n_runs):
        print(f'\nRun {n + 1} with seed: {seeds[n]}')
        set_seeds(int(seeds[n]))
        # generate an instance of the model
        m = QML(args.model_type, n)
        # load dataset
        data = np.load('../data/data_mfpt_3_classes.npz')
        # transform the dataset
        x_train, x_test, y_train,y_test, y_train_hot, y_test_hot = m.prepare_classic_dataset(data)
        x_train,x_test = m.prepare_quantum_dataset(x_train, x_test)
        data_loader = torch.utils.data.DataLoader(
            list(zip(x_train, y_train_hot)), batch_size=batch_size, shuffle=True, drop_last=True
        )
        # train the model
        h,acc = m.train(x_train, y_train, data_loader,args.learning_rate, args.epochs,batch_size)
        # save figures for evolution curves
        hist = np.vstack([t.detach().numpy() for t in h])  
        # evaluate model and generate dictionary
        d = m.evaluate(x_train, x_test, y_train, y_test)
        # generate pd.DataFrame with results
        d.pop('cm')
        results.append(d)
    end = timer()
    print('\nTime to run:  ',end - start)
    df = pd.DataFrame(data=results)
    print('\n\n\nRaw results:\n', df.to_string())
    print('\n\n\nAverage results:\n', df.describe().to_string())
```

**Comment by CatalinaAlbornoz** [comment]
Hi @henryzou50 ,

Thank you for opening this issue and sharing the example code.

My colleague Lee investigated and there seem to be two main causes for the difference in performance between `qiskit.aer` and `default.qubit`. 

On one hand it's almost inevitable for `default.qubit` to be faster. It has been developed and optimized to be very efficient in working with PennyLane, so it's generally expected that there will be a certain slowdown when using other simulators that haven't been optimized to work specifically with PennyLane.

That being said, it looks like `TorchLayer` likely has a for-loop that might be causing part of the slowdown in this case.

If you'd like to make a PR to work on making Qiskit circuits run faster that would be great. If you do make the PR please link to this issue so that the person reviewing the PR has the full context.

Let me know if you have any further questions or suggestions. We value this feedback very much and are always looking for ways to improve.

**Comment by CatalinaAlbornoz** [comment]
Hi @henryzou50 ,

I looked into this again and realized that part of the slowdown probably doesn't come from Torch at all. When you use `qiskit.aer` or `qiskit.remote`, gradients are only available via the parameter-shift rule. This is unavoidable since they act as a black box to PennyLane, and it's the only exact method that's compatible with actual quantum hardware. Unfortunately computing gradients with parameter-shift is much slower than other methods such as backprop or adjoint differentiation, as can bee seen in our [Adjoint differentiation demo](https://pennylane.ai/qml/demos/tutorial_adjoint_diff#performance). 

I ran a test with:
1. default.qubit
2. default.qubit with parameter-shift
3. qiskit.aer

The average training time was respectively:
1. 2s
2. 12s
3. 137s

This tells us two things: 
a. Having a slowdown of one order of magnitude is unavoidable given that we're restricted to using parameter-shift. 
b. There's one extra order of magnitude slowdown that can potentially be attributed to the conversion from PennyLane to Qiskit and Qiskit to PennyLane.

It's inevitable for there to be _some_ slowdown given that conversions need to be made, but maybe this can be the focus for improvement. We can test to see if this conversion is present when using the JAX interface too, to at least identify if this is generic to the plugin or specific to Torch.

Let us know if you have any insights pr findings on your side. We'll let you know what else we can find.

Note:
I based my test on our [Torch Layers demo](https://pennylane.ai/qml/demos/tutorial_qnn_module_torch). I'm adding my code below. Note that when using `parameter-shift` you need to expand the broadcasting with `qml.transforms.broadcast_expand` as shown in the code. I hope this can help you and your clients too!

```
import torch
import matplotlib.pyplot as plt
import numpy as np
from sklearn.datasets import make_moons
import pennylane as qml
from timeit import default_timer as timer


# Fix the dataset and problem
# ---------------------------

# Set random seeds
torch.manual_seed(42)
np.random.seed(42)
n_samples = 50

X, y = make_moons(n_samples=n_samples, noise=0.1)
y_ = torch.unsqueeze(torch.tensor(y), 1)  # used for one-hot encoded labels
y_hot = torch.scatter(torch.zeros((n_samples, 2)), 1, y_, 1)

c = ["#1f77b4" if y_ == 0 else "#ff7f0e" for y_ in y]  # colours for each class
plt.axis("off")
plt.scatter(X[:, 0], X[:, 1], c=c)
plt.show()

# Define the QNodes
# -----------------

n_qubits = 2

# Create the devices
dev_dq = qml.device("default.qubit", wires=n_qubits)
dev_qa = qml.device('qiskit.aer',wires=n_qubits,backend='unitary_simulator')

# Create a quantum function
def circuit(inputs, weights):
    qml.AngleEmbedding(inputs, wires=range(n_qubits))
    qml.BasicEntanglerLayers(weights, wires=range(n_qubits))
    return [qml.expval(qml.PauliZ(wires=i)) for i in range(n_qubits)]

# Create a QNode with default.qubit
qnode_dq = qml.QNode(circuit,dev_dq)

# Create a QNode with default.qubit and diff_method="parameter-shift"
qnode_dq_ps = qml.QNode(circuit,dev_dq, diff_method="parameter-shift")
expanded_qnode_dq_ps = qml.transforms.broadcast_expand(qnode_dq_ps)

# Create a QNode with qiskit.aer
qnode_qa = qml.QNode(circuit,dev_qa)
expanded_qnode_qa = qml.transforms.broadcast_expand(qnode_qa)

# Interface with Torch
# --------------------

n_layers = 6
weight_shapes = {"weights": (n_layers, n_qubits)}

def TorchModel(qnode):
    qlayer = qml.qnn.TorchLayer(qnode, weight_shapes)
    clayer_1 = torch.nn.Linear(2, 2)
    clayer_2 = torch.nn.Linear(2, 2)
    softmax = torch.nn.Softmax(dim=1)
    layers = [clayer_1, qlayer, clayer_2, softmax]
    model = torch.nn.Sequential(*layers)
    return model

# Create a model with default.qubit
model_dq = TorchModel(qnode_dq)
# Create a model with default.qubit and diff_method="parameter-shift"
model_dq_ps = TorchModel(expanded_qnode_dq_ps)
# Create a model with qiskit.aer
model_qa = TorchModel(expanded_qnode_qa)

# Build the optimization
# ----------------------

loss = torch.nn.L1Loss()

X = torch.tensor(X, requires_grad=True).float()
y_hot = y_hot.float()

batch_size = 5
batches = n_samples // batch_size

data_loader = torch.utils.data.DataLoader(
    list(zip(X, y_hot)), batch_size=5, shuffle=True, drop_last=True
)

epochs = 2
def optimization(model):
    opt = torch.optim.SGD(model.parameters(), lr=0.2)

    for epoch in range(epochs):

        running_loss = 0

        for xs, ys in data_loader:
            opt.zero_grad()

            loss_evaluated = loss(model(xs), ys)
            loss_evaluated.backward()

            opt.step()

            running_loss += loss_evaluated

        avg_loss = running_loss / batches
        print("Average loss over epoch {}: {:.4f}".format(epoch + 1, avg_loss))

# Get average training time per model
# -----------------------------------

num_runs = 2

# Obtain an average training time with default.qubit
timing_dq = []
for t in range(num_runs):
    start = timer()
    op = optimization(model_dq)
    end = timer()
    timing_dq.append(end - start)

print("timing_dq: ", np.mean(timing_dq))

# Obtain an average training time with default.qubit and diff_method="parameter-shift"
timing_dq_ps = []
for t in range(num_runs):
    start = timer()
    op = optimization(model_dq_ps)
    end = timer()
    timing_dq_ps.append(end - start)

print("timing_dq_ps: ", np.mean(timing_dq_ps))

# Obtain an average training time with qiskit.aer
timing_qa = []
for t in range(num_runs):
    start = timer()
    op = optimization(model_qa)
    end = timer()
    timing_qa.append(end - start)

print("timing_qa: ", np.mean(timing_qa))
```
